File: go/builder_pattern.go
package main

import (
	"errors"
	"fmt"
)

type Request struct {
	requestID string
	userID    int
	keyword   string
	sessionID int
	mode      string
}

type RequestBuilder struct {
	requestID string
	userID    int
	keyword   string
	sessionID int
	mode      string
}

func NewRequestBuilder() *RequestBuilder {
	return &RequestBuilder{}
}

func (rb *RequestBuilder) SetRequestID(requestID string) *RequestBuilder {
	rb.requestID = requestID
	return rb
}

func (rb *RequestBuilder) SetUserID(userID int) *RequestBuilder {
	rb.userID = userID
	return rb
}

func (rb *RequestBuilder) SetKeyword(keyword string) *RequestBuilder {
	rb.keyword = keyword
	return rb
}

func (rb *RequestBuilder) SetSessionID(sessionID int) *RequestBuilder {
	rb.sessionID = sessionID
	return rb
}

func (rb *RequestBuilder) SetMode(mode string) *RequestBuilder {
	rb.mode = mode
	return rb
}

// Build creates a new Request instance each time it's called
func (rb *RequestBuilder) Build() Request {
	return Request{
		requestID: rb.requestID,
		userID:    rb.userID,
		keyword:   rb.keyword,
		sessionID: rb.sessionID,
		mode:      rb.mode,
	}
}

// BuildWithValidation creates a Request with validation
func (rb *RequestBuilder) BuildWithValidation() (Request, error) {
	if rb.requestID == "" {
		return Request{}, errors.New("requestID is required")
	}
	return rb.Build(), nil
}

// Reset clears the builder for reuse
func (rb *RequestBuilder) Reset() *RequestBuilder {
	rb.requestID = ""
	rb.userID = 0
	rb.keyword = ""
	rb.sessionID = 0
	rb.mode = ""
	return rb
}

func main() {
	// Basic usage
	request := NewRequestBuilder().SetRequestID("request1").Build()
	fmt.Printf("request: %+v\n", request)

	request = NewRequestBuilder().SetMode("autosuggest").SetKeyword("query").Build()
	fmt.Printf("request: %+v\n", request)

	// Demonstrating the fix - builder reuse now works correctly
	builder := NewRequestBuilder()
	req1 := builder.SetRequestID("request1").SetUserID(100).Build()
	req2 := builder.SetRequestID("request2").SetUserID(200).Build()

	fmt.Printf("req1: %+v\n", req1) // Will have requestID="request1", userID=100
	fmt.Printf("req2: %+v\n", req2) // Will have requestID="request2", userID=200

	// Using validation
	req3, err := NewRequestBuilder().SetMode("search").BuildWithValidation()
	if err != nil {
		fmt.Printf("Validation error: %v\n", err)
	} else {
		fmt.Printf("req3: %+v\n", req3)
	}

	// Using reset for builder reuse
	builder.Reset().SetRequestID("request3").SetMode("search")
	req4 := builder.Build()
	fmt.Printf("req4: %+v\n", req4)
}

File: java/builder_pattern/BuilderPattern.java
final class Request {
    private final String requestID;
    private final int userID;
    private final String keyword;
    private final int sessionID;
    private final String mode;

    Request(String requestID, int userID, String keyword, int sessionID, String mode) {
        this.requestID = requestID;
        this.userID = userID;
        this.keyword = keyword;
        this.sessionID = sessionID;
        this.mode = mode;
    }

    @Override
    public String toString() {
        return "{requestID=" + requestID
                + ", userID=" + userID
                + ", keyword=" + keyword
                + ", sessionID=" + sessionID
                + ", mode=" + mode + "}";
    }
}

class RequestBuilder {
    private String requestID = "";
    private int userID = 0;
    private String keyword = "";
    private int sessionID = 0;
    private String mode = "";

    public RequestBuilder setRequestID(String requestID) {
        this.requestID = requestID;
        return this;
    }

    public RequestBuilder setUserID(int userID) {
        this.userID = userID;
        return this;
    }

    public RequestBuilder setKeyword(String keyword) {
        this.keyword = keyword;
        return this;
    }

    public RequestBuilder setSessionID(int sessionID) {
        this.sessionID = sessionID;
        return this;
    }

    public RequestBuilder setMode(String mode) {
        this.mode = mode;
        return this;
    }

    public Request build() {
        return new Request(requestID, userID, keyword, sessionID, mode);
    }

    public Request buildWithValidation() {
        if (requestID == null || requestID.isEmpty()) {
            throw new IllegalStateException("requestID is required");
        }
        return build();
    }

    public RequestBuilder reset() {
        this.requestID = "";
        this.userID = 0;
        this.keyword = "";
        this.sessionID = 0;
        this.mode = "";
        return this;
    }
}

public class BuilderPattern {
    public static void main(String[] args) {
        Request request = new RequestBuilder().setRequestID("request1").build();
        System.out.println("request: " + request);

        request = new RequestBuilder().setMode("autosuggest").setKeyword("query").build();
        System.out.println("request: " + request);

        RequestBuilder builder = new RequestBuilder();
        Request req1 = builder.setRequestID("request1").setUserID(100).build();
        Request req2 = builder.setRequestID("request2").setUserID(200).build();

        System.out.println("req1: " + req1);
        System.out.println("req2: " + req2);

        try {
            Request req3 = new RequestBuilder().setMode("search").buildWithValidation();
            System.out.println("req3: " + req3);
        } catch (IllegalStateException e) {
            System.out.println("Validation error: " + e.getMessage());
        }

        builder.reset().setRequestID("request3").setMode("search");
        Request req4 = builder.build();
        System.out.println("req4: " + req4);
    }
}
